extensions/oplogPopulator/modules/ConnectorsManager.js

```diff
@@ -235,19 +255,23 @@ class ConnectorsManager {
                 this._metricsHandler.onConnectorDestroyed();
                 this._logger.info('Successfully destroyed a connector', {
                     method: 'ConnectorsManager._spawnOrDestroyConnector',
-                    connector: connector.name
+                    connector: connector.name,
                 });
                 return true;
             } else if (!connector.isRunning && connector.bucketCount > 0) {
                 await connector.spawn();
                 this._metricsHandler.onConnectorsInstantiated(false);
                 this._logger.info('Successfully spawned a connector', {
                     method: 'ConnectorsManager._spawnOrDestroyConnector',
-                    connector: connector.name
+                    connector: connector.name,
                 });
                 return true;
-            } else if (connector.isRunning) {
-                return connector.updatePipeline(true);
+            } else if (connector.isRunning && this._allocationStrategy.canUpdate(connector)) {
+                const isConnectorUpdated = connector.updatePipeline(true);
+                if (isConnectorUpdated) {
+                    this._allocationStrategy.onConnectorUpdated(connector);
```

Comment:
this is kind of a stretch : `onBucketRemoved` and `onConnectorUpdated` is not actually part of `AllocationStrategy` interface, but only of `RetainBucketsDecorator` implementation.

It works only because this is Javascript (would not build with typescript), and because RetainBucketsDecorator is the "first" strategy in the chain (i.e. the abstraction is broken, and we have coupling between the "choice" of strategy and its "users")

So need either:
* Apply inversion of control: e.g. Allocator and ConnectorsManager emit an event or allow registering a callback, so that `RetainBucketsDecorator` can register for these 'events'
* Extend the AllocationStrategy interface to have the extra events (even though it seems a bit of a stretch to make them part of the interface)
* Introduce one (or two) interfaces for this/these 2 events. This interface would be implemented by `RetainBucketsDecorator` only (as today), the only difference would be that ConnectorsManager and Allocator would get 2 objects: one strategy and the object having these interfaces... (in typepscript, we could materialize with multiple interfaces; in other languages with multiple inheritance)
* Similar to the last approach above, these classes could take `onConnectorUpdated` and `onBucketRemoved` callbacks in their constructor, which would be implemented by binding them to the `RetainBucketsDecorator` methods...

Reply:
As discussed, I chose to bind the events from the classes from the oplog populator, as we know there that we use the RetainBucket strategy (may change someday, but for now, we only do that).
I chose events because it allows the not alter the classes: they will just emit events.